third_party/move/move-compiler-v2/src/pipeline/ability_processor.rs

```diff
@@ -0,0 +1,656 @@
+        let copy_drop =
+            analyzer.state_per_instruction_with_default(state_map, &code, &cfg, |_, after| {
+                after.clone()
+            });
+
+        // Run instrumenter
+        let mut instrumenter = Instrumenter {
+            builder,
+            live_var,
+            copy_drop,
+        };
+        instrumenter.run(code);
+        instrumenter.builder.data
+    }
+
+    fn name(&self) -> String {
+        "AbilityProcessor".to_owned()
+    }
+}
+
+// =================================================================================================
+// Copy/Drop Analysis
+
+#[derive(AbstractDomain, Debug, Clone, Default)]
+struct CopyDropState {
+    needs_copy: SetDomain<TempIndex>,
+    needs_drop: SetDomain<TempIndex>,
+    moved: SetDomain<TempIndex>,
+}
+
+struct CopyDropAnalysis<'a> {
+    target: &'a FunctionTarget<'a>,
+    live_var: &'a LiveVarAnnotation,
+    lifetime: &'a LifetimeAnnotation,
+    exit_state: &'a ExitStateAnnotation,
+}
+
+impl<'a> DataflowAnalysis for CopyDropAnalysis<'a> {}
+
+impl<'a> TransferFunctions for CopyDropAnalysis<'a> {
+    type State = CopyDropState;
+
+    const BACKWARD: bool = false;
+
+    fn execute(&self, state: &mut Self::State, instr: &Bytecode, offset: CodeOffset) {
+        use Bytecode::*;
+        // Clear local state info
+        state.needs_copy.clear();
+        state.needs_drop.clear();
+        let live_var = self.live_var.get_info_at(offset);
+        let lifetime = self.lifetime.get_info_at(offset);
+        let exit_state = self.exit_state.get_state_at(offset);
+        // Only non-primitive types need a copy
+        let type_needs_copy = |temp: &TempIndex| {
+            let ty = self.target.get_local_type(*temp);
+            !ty.is_primitive()
+        };
+        // Only temps which are used after or borrowed need a copy
+        let temp_needs_copy =
+            |temp| live_var.after.contains_key(temp) || lifetime.before.is_borrowed(*temp);
+        // References always need to be dropped to satisfy bytecode verifier borrow analysis, other values
+        // only if this execution path can return.
+        let temp_needs_drop = |temp: &TempIndex| {
+            self.target.get_local_type(*temp).is_reference() || exit_state.may_return()
+        };
+        match instr {
+            Assign(_, _, src, AssignKind::Inferred) => {
+                if temp_needs_copy(src) {
+                    state.needs_copy.insert(*src);
+                } else {
+                    state.moved.insert(*src);
+                }
+            },
+            Assign(_, _, src, AssignKind::Move) => {
+                state.moved.insert(*src);
+            },
+            Call(_, _, _, srcs, ..) => {
+                // For arguments, we also need to check the case that a src is not used after this program point
+                // but again in the argument list. Also, in difference to assign, we only need to copy the
```

Comment:
I thought this was referring to the code: `temp_needs_copy(src) || srcs[i + 1..].contains(src)`. I suggested "or" because of this. Re-reading this, I think both would be fine. Ok to leave this as is.

Reply:
its the 'not used after but...' where you scanned 'used after or ..'. I changed it to avoid the confusion.